cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/workflow/BasicWorkflowToken.java

```diff
@@ -27,19 +32,139 @@
  */
 public class BasicWorkflowToken implements WorkflowToken {
   private Map<String, Map<String, Long>> mapReduceCounters;
+  private Map<String, List<NodeValueEntry>> tokenValueMap = Maps.newHashMap();
```

Comment:
Can this be final? Instead of the setter method, you can have a replace method that will clear the map and replace the content.